cmd/authgear/elasticsearch/reindex.go

```diff
@@ -126,69 +174,128 @@ func (q *Reindexer) Reindex(es *elasticsearch.Client) (err error) {
 		after = items[len(items)-1].Cursor
 
 		// Process the items
-		buf := &bytes.Buffer{}
 		for _, item := range items {
 			user := item.Value.(*model.ElasticsearchUserRaw)
-			fmt.Printf("Indexing app (%s) user (%s)\n", user.AppID, user.ID)
-			err = q.writeBody(buf, user)
+			source := libes.RawToSource(user)
+			allUserIDs[user.ID] = struct{}{}
+
+			var bodyBytes []byte
+			bodyBytes, err = json.Marshal(source)
 			if err != nil {
 				return
 			}
-		}
 
-		var res *esapi.Response
-		res, err = es.Bulk(buf, func(o *esapi.BulkRequest) {
-			o.Index = libes.IndexNameUser
-		})
-		if err != nil {
-			return
-		}
-		defer res.Body.Close()
-		if res.IsError() {
-			err = fmt.Errorf("%v", res)
-			return
+			err = bulkIndexer.Add(
+				ctx,
+				esutil.BulkIndexerItem{
+					Action:     "index",
+					DocumentID: fmt.Sprintf("%s:%s", source.AppID, source.ID),
+					Body:       bytes.NewReader(bodyBytes),
+					OnFailure: func(
+						ctx context.Context,
+						item esutil.BulkIndexerItem,
+						res esutil.BulkIndexerResponseItem,
+						err error,
+					) {
+						if err != nil {
+							fmt.Fprintf(os.Stderr, "%v\n", err)
+						} else {
+							fmt.Fprintf(os.Stderr, "%v: %v\n", res.Error.Type, res.Error.Reason)
+						}
+					},
+				},
+			)
+			if err != nil {
+				return
+			}
 		}
 	}
 
-	return nil
+	return allUserIDs, nil
 }
 
-func (q *Reindexer) writeBody(buf io.Writer, raw *model.ElasticsearchUserRaw) (err error) {
-	source := libes.RawToSource(raw)
-	id := fmt.Sprintf("%s:%s", source.AppID, source.ID)
-	action := map[string]interface{}{
-		"index": map[string]interface{}{
-			"_id": id,
-		},
-	}
-	actionBytes, err := json.Marshal(action)
-	if err != nil {
-		return
-	}
+func (q *Reindexer) cleanupDeletedUsers(ctx context.Context, es *elasticsearch.Client, bulkIndexer esutil.BulkIndexer, userIDs map[string]struct{}) error {
+	// Each user ID is 128-bit UUID == 16 bytes
+	// 1M users take 16MiB memory
+	size := 1000
+	var searchAfter interface{}
 
-	_, err = buf.Write(actionBytes)
-	if err != nil {
-		return
-	}
+	underscoreIDsToDelete := make(map[string]struct{})
+	for {
+		bodyJSONValue := map[string]interface{}{
```

Comment:
I think we need to add app_id to the search?

Reply:
Fixed.